// js-binary-schema-parser/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../../_virtual/_commonjsHelpers.js');
var index = require('../../../_virtual/index3.js');

"use strict";
Object.defineProperty(index.__exports, "__esModule", {
  value: true
});
exports.loop = index.__exports.loop = exports.conditional = index.__exports.conditional = exports.parse = index.__exports.parse = void 0;
var parse = function parse2(stream, schema) {
  var result = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
  var parent = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : result;
  if (Array.isArray(schema)) {
    schema.forEach(function(partSchema) {
      return parse2(stream, partSchema, result, parent);
    });
  } else if (typeof schema === "function") {
    schema(stream, result, parent, parse2);
  } else {
    var key = Object.keys(schema)[0];
    if (Array.isArray(schema[key])) {
      parent[key] = {};
      parse2(stream, schema[key], result, parent[key]);
    } else {
      parent[key] = schema[key](stream, result, parent, parse2);
    }
  }
  return result;
};
exports.parse = index.__exports.parse = parse;
var conditional = function conditional2(schema, conditionFunc) {
  return function(stream, result, parent, parse3) {
    if (conditionFunc(stream, result, parent)) {
      parse3(stream, schema, result, parent);
    }
  };
};
exports.conditional = index.__exports.conditional = conditional;
var loop = function loop2(schema, continueFunc) {
  return function(stream, result, parent, parse3) {
    var arr = [];
    var lastStreamPos = stream.pos;
    while (continueFunc(stream, result, parent)) {
      var newParent = {};
      parse3(stream, schema, result, newParent);
      if (stream.pos === lastStreamPos) {
        break;
      }
      lastStreamPos = stream.pos;
      arr.push(newParent);
    }
    return arr;
  };
};
exports.loop = index.__exports.loop = loop;

exports.default = index.__exports;
