// pixi-filters/lib/advanced-bloom/ExtractBrightnessFilter.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pixi_js = require('pixi.js');
var _default$1 = require('../defaults/default2.js');
var _default = require('../defaults/default.js');
var extractBrightness$1 = require('./extract-brightness2.js');
var extractBrightness = require('./extract-brightness.js');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
const _ExtractBrightnessFilter = class _ExtractBrightnessFilter extends pixi_js.Filter {
  constructor(options) {
    options = { ..._ExtractBrightnessFilter.DEFAULT_OPTIONS, ...options };
    const gpuProgram = pixi_js.GpuProgram.from({
      vertex: {
        source: _default["default"],
        entryPoint: "mainVertex"
      },
      fragment: {
        source: extractBrightness["default"],
        entryPoint: "mainFragment"
      }
    });
    const glProgram = pixi_js.GlProgram.from({
      vertex: _default$1["default"],
      fragment: extractBrightness$1["default"],
      name: "extract-brightness-filter"
    });
    super({
      gpuProgram,
      glProgram,
      resources: {
        extractBrightnessUniforms: {
          uThreshold: { value: options.threshold, type: "f32" }
        }
      }
    });
    __publicField(this, "uniforms");
    this.uniforms = this.resources.extractBrightnessUniforms.uniforms;
  }
  /**
   * Defines how bright a color needs to be extracted.
   * @default 0.5
   */
  get threshold() {
    return this.uniforms.uThreshold;
  }
  set threshold(value) {
    this.uniforms.uThreshold = value;
  }
};
/** Default values for options. */
__publicField(_ExtractBrightnessFilter, "DEFAULT_OPTIONS", {
  threshold: 0.5
});
let ExtractBrightnessFilter = _ExtractBrightnessFilter;

exports.ExtractBrightnessFilter = ExtractBrightnessFilter;
