// pixi-filters/lib/backdrop-blur/BackdropBlurFilter.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pixi_js = require('pixi.js');
var _default$1 = require('../defaults/default2.js');
var _default = require('../defaults/default.js');
var backdropBlurBlend$1 = require('./backdrop-blur-blend.js');
var backdropBlurBlend = require('./backdrop-blur-blend2.js');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class BackdropBlurFilter extends pixi_js.BlurFilter {
  /**
   * @param options - The options of the blur filter.
   */
  constructor(options) {
    super(options);
    __publicField(this, "_blendPass");
    this.blendRequired = true;
    this.padding = 0;
    this._blendPass = new pixi_js.Filter({
      gpuProgram: pixi_js.GpuProgram.from({
        vertex: {
          source: _default["default"],
          entryPoint: "mainVertex"
        },
        fragment: {
          source: backdropBlurBlend["default"],
          entryPoint: "mainFragment"
        }
      }),
      glProgram: pixi_js.GlProgram.from({
        vertex: _default$1["default"],
        fragment: backdropBlurBlend$1["default"],
        name: "drop-shadow-filter"
      }),
      resources: {
        uBackground: pixi_js.Texture.EMPTY
      }
    });
  }
  /**
   * Override existing apply method in `Filter`
   * @override
   * @ignore
   */
  apply(filterManager, input, output, clearMode) {
    const backTexture = filterManager._activeFilterData.backTexture;
    const blurredBackground = pixi_js.TexturePool.getSameSizeTexture(input);
    super.apply(filterManager, backTexture, blurredBackground, true);
    this._blendPass.resources.uBackground = blurredBackground.source;
    this._blendPass.apply(filterManager, input, output, clearMode);
    pixi_js.TexturePool.returnTexture(blurredBackground);
  }
  updatePadding() {
    this.padding = 0;
  }
}

exports.BackdropBlurFilter = BackdropBlurFilter;
