// js-binary-schema-parser/lib/parsers/uint8.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('../../../../_virtual/_commonjsHelpers.js');
var uint8 = require('../../../../_virtual/uint82.js');

"use strict";
Object.defineProperty(uint8.__exports, "__esModule", {
  value: true
});
exports.readBits = uint8.__exports.readBits = exports.readArray = uint8.__exports.readArray = exports.readUnsigned = uint8.__exports.readUnsigned = exports.readString = uint8.__exports.readString = exports.peekBytes = uint8.__exports.peekBytes = exports.readBytes = uint8.__exports.readBytes = exports.peekByte = uint8.__exports.peekByte = exports.readByte = uint8.__exports.readByte = exports.buildStream = uint8.__exports.buildStream = void 0;
var buildStream = function buildStream2(uint8Data) {
  return {
    data: uint8Data,
    pos: 0
  };
};
exports.buildStream = uint8.__exports.buildStream = buildStream;
var readByte = function readByte2() {
  return function(stream) {
    return stream.data[stream.pos++];
  };
};
exports.readByte = uint8.__exports.readByte = readByte;
var peekByte = function peekByte2() {
  var offset = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : 0;
  return function(stream) {
    return stream.data[stream.pos + offset];
  };
};
exports.peekByte = uint8.__exports.peekByte = peekByte;
var readBytes = function readBytes2(length) {
  return function(stream) {
    return stream.data.subarray(stream.pos, stream.pos += length);
  };
};
exports.readBytes = uint8.__exports.readBytes = readBytes;
var peekBytes = function peekBytes2(length) {
  return function(stream) {
    return stream.data.subarray(stream.pos, stream.pos + length);
  };
};
exports.peekBytes = uint8.__exports.peekBytes = peekBytes;
var readString = function readString2(length) {
  return function(stream) {
    return Array.from(readBytes(length)(stream)).map(function(value) {
      return String.fromCharCode(value);
    }).join("");
  };
};
exports.readString = uint8.__exports.readString = readString;
var readUnsigned = function readUnsigned2(littleEndian) {
  return function(stream) {
    var bytes = readBytes(2)(stream);
    return littleEndian ? (bytes[1] << 8) + bytes[0] : (bytes[0] << 8) + bytes[1];
  };
};
exports.readUnsigned = uint8.__exports.readUnsigned = readUnsigned;
var readArray = function readArray2(byteSize, totalOrFunc) {
  return function(stream, result, parent) {
    var total = typeof totalOrFunc === "function" ? totalOrFunc(stream, result, parent) : totalOrFunc;
    var parser = readBytes(byteSize);
    var arr = new Array(total);
    for (var i = 0; i < total; i++) {
      arr[i] = parser(stream);
    }
    return arr;
  };
};
exports.readArray = uint8.__exports.readArray = readArray;
var subBitsTotal = function subBitsTotal2(bits, startIndex, length) {
  var result = 0;
  for (var i = 0; i < length; i++) {
    result += bits[startIndex + i] && Math.pow(2, length - i - 1);
  }
  return result;
};
var readBits = function readBits2(schema) {
  return function(stream) {
    var _byte = readByte()(stream);
    var bits = new Array(8);
    for (var i = 0; i < 8; i++) {
      bits[7 - i] = !!(_byte & 1 << i);
    }
    return Object.keys(schema).reduce(function(res, key) {
      var def = schema[key];
      if (def.length) {
        res[key] = subBitsTotal(bits, def.index, def.length);
      } else {
        res[key] = bits[def.index];
      }
      return res;
    }, {});
  };
};
exports.readBits = uint8.__exports.readBits = readBits;

exports.default = uint8.__exports;
